File: src/quest_editor/actions/RotateEntityAction.ts
import { Action } from "../../core/undo/Action";
import { QuestEntityModel } from "../model/QuestEntityModel";
import { entity_data } from "../../core/data_formats/parsing/quest/entities";
import { quest_editor_store } from "../stores/QuestEditorStore";
import { Euler } from "three";

export class RotateEntityAction implements Action {
    readonly description: string;

    constructor(
        private entity: QuestEntityModel,
        private old_rotation: Euler,
        private new_rotation: Euler,
        private world: boolean,
    ) {
        this.description = `Rotate ${entity_data(entity.type).name}`;
    }

    undo(): void {
        quest_editor_store.set_selected_entity(this.entity);

        if (this.world) {
            this.entity.set_world_rotation(this.old_rotation);
        } else {
            this.entity.set_rotation(this.old_rotation);
        }
    }

    redo(): void {
        quest_editor_store.set_selected_entity(this.entity);

        if (this.world) {
            this.entity.set_world_rotation(this.new_rotation);
        } else {
            this.entity.set_rotation(this.new_rotation);
        }
    }
}

File: web/src/main/kotlin/world/phantasmal/web/questEditor/actions/RotateEntityAction.kt
package world.phantasmal.web.questEditor.actions

import world.phantasmal.observable.change
import world.phantasmal.web.core.actions.Action
import world.phantasmal.web.externals.three.Euler
import world.phantasmal.web.questEditor.models.QuestEntityModel

class RotateEntityAction(
    private val setSelectedEntity: (QuestEntityModel<*, *>) -> Unit,
    private val entity: QuestEntityModel<*, *>,
    private val newRotation: Euler,
    private val oldRotation: Euler,
    private val world: Boolean,
) : Action {
    override val description: String = "Rotate ${entity.type.simpleName}"

    override fun execute() {
        change {
            setSelectedEntity(entity)

            if (world) {
                entity.setWorldRotation(newRotation)
            } else {
                entity.setRotation(newRotation)
            }
        }
    }

    override fun undo() {
        change {
            setSelectedEntity(entity)

            if (world) {
                entity.setWorldRotation(oldRotation)
            } else {
                entity.setRotation(oldRotation)
            }
        }
    }
}
